From stable+bounces-114301-greg=kroah.com@vger.kernel.org Fri Feb  7 20:28:59 2025
From: "Darrick J. Wong" <djwong@kernel.org>
Date: Fri, 07 Feb 2025 11:28:54 -0800
Subject: xfs: lock dquot buffer before detaching dquot from b_li_list
To: djwong@kernel.org, xfs-stable@lists.linux.dev
Cc: hch@lst.de, cem@kernel.org, stable@vger.kernel.org
Message-ID: <173895601566.3373740.7315790259628729484.stgit@frogsfrogsfrogs>

From: Darrick J. Wong <djwong@kernel.org>

commit 111d36d6278756128b7d7fab787fdcbf8221cd98 upstream

We have to lock the buffer before we can delete the dquot log item from
the buffer's log item list.

Cc: <stable@vger.kernel.org> # v6.13-rc3
Fixes: acc8f8628c3737 ("xfs: attach dquot buffer to dquot log item buffer")
Signed-off-by: "Darrick J. Wong" <djwong@kernel.org>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Carlos Maiolino <cem@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 fs/xfs/xfs_dquot.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)


--- a/fs/xfs/xfs_dquot.c
+++ b/fs/xfs/xfs_dquot.c
@@ -87,8 +87,9 @@ xfs_dquot_detach_buf(
 	}
 	spin_unlock(&qlip->qli_lock);
 	if (bp) {
+		xfs_buf_lock(bp);
 		list_del_init(&qlip->qli_item.li_bio_list);
-		xfs_buf_rele(bp);
+		xfs_buf_relse(bp);
 	}
 }
 
